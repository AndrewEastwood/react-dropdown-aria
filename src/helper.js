export function createStyleObject(width, height, maxHeight) {
  const style = {};
  if (width) {
    style.width = width;
  }
  if (height) {
    style.height = height;
  }
  if (maxHeight) {
    style.maxHeight = maxHeight;
    style.overflowY = 'scroll';
  }
  return style;
}

export const KEY_CODES = {
  UP_ARROW: 38,
  DOWN_ARROW: 40,
  ENTER: 13,
  ESCAPE: 27,
  TAB: 9,
};

export const NAVIGATION_KEYS = [KEY_CODES.ESCAPE, KEY_CODES.UP_ARROW, KEY_CODES.DOWN_ARROW];
