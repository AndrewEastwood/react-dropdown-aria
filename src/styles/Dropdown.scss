@import './Colours.scss';

.dropdown {
  position: relative;
  display: flex;
  flex-direction: column;
}

.dropdown-select {
  font-size: 1em;
  display: flex;
  flex-direction: row;
  align-items: center;
  width: 100%;
  height: 100%;
  background-color: lighten(lightGrey, 45%);
  padding: 9px 5px;
  margin: 0;
  border-radius: 0px;
  border-bottom: 2px solid lighten($grey, 20%);
  border-top: none;
  border-right: none;
  border-left: none;
  text-align: left;
  cursor: pointer;
  outline: none;

  &:hover, &:focus {
    box-shadow: 0px 1px 3px 2px lighten($grey, 35%);
  }

  &:disabled {
    cursor: not-allowed;
  }

  .displayed-value {
    flex: 1;
    border-right: 1px solid lighten($grey, 30%);
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
    color: $black;
  }

  .no-arrow {
    border-right: none;
  }

  .grey {
    color: $grey;
  }

  .center-text {
    text-align: center;
  }

  .dropdown-arrow {
    content: "";
    width: 0;
    height: 0;
    margin-left: 8px;
    margin-right: 5px;
    border-right: 5px solid transparent;
    border-left: 5px solid transparent;
  }

  .down {
    border-top: 5px solid $grey;
    border-bottom: 0;
  }

  .up {
    border-bottom: 5px solid $grey;
    border-top: 0;
  }
}

.dropdown-content {
  width: 100%;
  max-height: 175px;
  z-index: 9999;
  overflow-x: hidden;
  position: absolute;
  left: 0;
  list-style-type: none;
  margin: 0;
  padding: 2px 0;
  background-color: white;
  color: black;
  border-radius: 2px;
  display: none;
  box-sizing: border-box;

  &::-webkit-scrollbar {
    width: 5px;
  }

  &::-webkit-scrollbar-track {
    background: #ddd;
  }

  &::-webkit-scrollbar-thumb {
    background: #666;
  }
}

.dropdown-content-down {
  top: 100%;
  box-shadow: 0px 3px 3px 2px lighten($grey, 35%);
}

.dropdown-content-up {
  bottom: 105%;
  box-shadow: 0px -3px 3px 2px lighten($grey, 35%);
}

.dropdown-option, .dropdown-option-selected {
  font-size: 0.95em;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  padding: 5px 10px;
  width: 100%;
  text-align: left;
  cursor: pointer;
  outline: none;
  background-color: white;
  border: none;

  &:hover, &:focus {
    background-color: lighten($grey, 35%);
  }

  .option-icon {
    padding-right: 5px;
  }
}

.dropdown-option-selected {
  background-color: lighten($purple-light, 30%);
  &:hover, &:focus {
    background-color: lighten($purple-light, 20%);
  }
}

.dropdown-content-open {
  display: block;
}
