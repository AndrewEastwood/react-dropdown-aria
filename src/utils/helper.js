export const KEY_CODES = {
  UP_ARROW: 38,
  DOWN_ARROW: 40,
  ENTER: 13,
  ESCAPE: 27,
  TAB: 9,
  PAGE_UP: 33,
  PAGE_DOWN: 34,
};

export const NAVIGATION_KEYS = [KEY_CODES.ESCAPE, KEY_CODES.UP_ARROW, KEY_CODES.DOWN_ARROW, KEY_CODES.PAGE_UP, KEY_CODES.PAGE_DOWN];
