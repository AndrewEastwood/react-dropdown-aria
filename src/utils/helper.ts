import { Option, OptionGroup } from './types';

export function isOptionGroup(option: Option | OptionGroup): option is OptionGroup {
  return (option as OptionGroup).groupOptions !== undefined;
}

export function isArray(value: any | any[]): value is any[] {
  return Array.isArray(value);
}
