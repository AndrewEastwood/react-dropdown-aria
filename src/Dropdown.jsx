import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import defaultOptionRenderer from './utils/defaultOptionRenderer';
import { createStyleObject, KEY_CODES, NAVIGATION_KEYS } from './utils/helper';
import './styles/Dropdown.scss';

class Dropdown extends Component {
  constructor(props) {
    super(props);
    this.state = {
      open: false,
      searchTerm: '',
      searchTimer: -1,
      focusedIndex: -1,
      internalSelectedOption: null,
    };
    this.elements = [];
  }

  onBlur = (e) => {
    const target = e.nativeEvent.relatedTarget;
    if (!target) {
      this.setState({ open: false });
      return;
    }
    const classValue = target.classList.value;
    if (classValue.indexOf('dropdown') === -1) {
      this.setState({ open: false });
    }
  }

  onDropdownClick = () => {
    const { disabled } = this.props;
    if (!disabled) this.setState(p => ({ open: !p.open }));
  }

  onOptionClicked = ({ nativeEvent }) => {
    const { setSelected } = this.props;
    const selectedOption = nativeEvent.target.innerText;
    setSelected(selectedOption);
    this.setState(p => ({ open: !p.open, internalSelectedOption: selectedOption }));
  }

  onKeyDown = ({ nativeEvent }) => {
    const key = nativeEvent.key.toLowerCase();
    const { keyCode } = nativeEvent;
    const { searchable } = this.props;

    if (NAVIGATION_KEYS.includes(keyCode)) {
      nativeEvent.preventDefault();
      this.onNavigation(keyCode);
    } else if (key.length === 1 && searchable) {
      this.searchDropdown(key);
    }
  }

  onNavigation = (keyCode) => {
    const { focusedIndex } = this.state;
    switch (keyCode) {
      case KEY_CODES.UP_ARROW:
        if (focusedIndex === -1) {
          this.setState({ focusedIndex: 0 }, this.setFocus);
        } else if (focusedIndex === 0) {
          this.setState({ focusedIndex: this.elements.length - 1 }, this.setFocus);
        } else {
          this.setState(p => ({ focusedIndex: p.focusedIndex - 1 }), this.setFocus);
        }
        break;
      case KEY_CODES.DOWN_ARROW:
        if (focusedIndex === -1 || focusedIndex === this.elements.length - 1) {
          this.setState({ focusedIndex: 0 }, this.setFocus);
        } else {
          this.setState(p => ({ focusedIndex: p.focusedIndex + 1 }), this.setFocus);
        }
        break;
      case KEY_CODES.ESCAPE:
        this.setState({ open: false });
        this.button.focus();
        break;
      default:
        break;
    }
  }

  setFocus = () => {
    const { focusedIndex } = this.state;
    this.elements[focusedIndex].focus();
  }

  searchDropdown = (key) => {
    const { searchTerm } = this.state;
    const oldTerm = searchTerm;
    this.setState(p => ({ searchTerm: p.searchTerm + key }));
    this.searchList(oldTerm + key);

    this.clearTimer();
    const timer = setTimeout(this.clearSearch, 1500);
    this.setState({ searchTimer: timer });
  }

  clearTimer = () => {
    const { searchTimer } = this.state;
    if (searchTimer !== -1) {
      clearTimeout(searchTimer);
      this.setState({ searchTimer: -1 });
    }
  }

  clearSearch = () => {
    this.setState({ searchTerm: '' });
  }

  searchList = (value) => {
    const element = this.elements.find(el => el.innerText.toLowerCase().indexOf(value) === 0);
    if (element) element.focus();
  }

  renderOptions = () => {
    const { optionRenderer, selectedOption, options } = this.props;
    const { internalSelectedOption } = this.state;

    this.elements = []; // Reset reference array

    if (optionRenderer) {
      return optionRenderer(selectedOption || internalSelectedOption, options, this.onOptionClicked, this.elements);
    }

    return defaultOptionRenderer(selectedOption || internalSelectedOption, options, this.onOptionClicked, this.elements);
  }

  render() {
    // Please Keep Alphabetical
    const { ariaDescribedBy, ariaLabel, ariaLabelledBy, centerText, className, disabled, height, hideArrow, id, maxContentHeight, placeholder, selectedOption, width } = this.props;
    const { internalSelectedOption, open } = this.state;

    const displayedValue = selectedOption || internalSelectedOption || placeholder || '';
    const dropdownButtonClass = classNames('dropdown-select', className);
    const displayedValueClass = classNames('displayed-value', { grey: !selectedOption && !internalSelectedOption, 'no-arrow': hideArrow, 'center-text': centerText });
    const contentClass = classNames('dropdown-content', { 'dropdown-content-open': open });
    const arrowClass = open ? 'dropdown-arrow up' : 'dropdown-arrow down';
    const listStyle = maxContentHeight ? { maxHeight: maxContentHeight, overflowY: 'scroll' } : {};

    return (
      <div className="dropdown" onBlur={this.onBlur} onKeyDown={this.onKeyDown} style={createStyleObject(width, height)}>
        <button
          aria-label={ariaLabel}
          ari-describedby={ariaDescribedBy}
          aria-labelledby={ariaLabelledBy}
          className={dropdownButtonClass}
          disabled={disabled}
          id={id}
          onClick={this.onDropdownClick}
          ref={btn => this.button = btn}
          type="button"
        >
          <div className={displayedValueClass}>{ displayedValue }</div>
          { !hideArrow && <div className={arrowClass} /> }
        </button>
        <ul className={contentClass} style={listStyle}>{ this.renderOptions() }</ul>
      </div>
    );
  }
}

// Please Keep Alphabetical
Dropdown.propTypes = {
  ariaDescribedBy: PropTypes.string,
  ariaLabel: PropTypes.string,
  ariaLabelledBy: PropTypes.string,
  className: PropTypes.string,
  centerText: PropTypes.bool,
  disabled: PropTypes.bool,
  height: PropTypes.number,
  hideArrow: PropTypes.bool,
  id: PropTypes.string,
  optionRenderer: PropTypes.func,
  maxContentHeight: PropTypes.number,
  options: PropTypes.array.isRequired,
  placeholder: PropTypes.string,
  searchable: PropTypes.bool,
  selectedOption: PropTypes.string,
  setSelected: PropTypes.func.isRequired,
  width: PropTypes.number,
};

// Please Keep Alphabetical
Dropdown.defaultProps = {
  ariaDescribedBy: null,
  ariaLabel: null,
  ariaLabelledBy: null,
  className: undefined,
  centerText: false,
  disabled: false,
  height: null,
  hideArrow: false,
  id: undefined,
  optionRenderer: undefined,
  maxContentHeight: null,
  placeholder: 'Select ...',
  searchable: true,
  selectedOption: null,
  width: null,
};

export default Dropdown;
