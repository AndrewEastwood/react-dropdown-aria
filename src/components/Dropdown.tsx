import React, { Component, MouseEvent, KeyboardEvent, createRef } from 'react';
import { css, cx } from 'emotion';
import { KEY_CODES, NAVIGATION_KEYS } from '../utils/helper';
import defaultOptionRenderer from '../utils/defaultOptionRenderer';
import defaultStyles from '../styles/Dropdown';
import { DropdownOption, OptionGroup, Option } from './OptionItem';

type StyleFunction = (base: {}, state: DropdownState, extraState?: {}) => {};

interface DropdownStyle {
  arrow?: StyleFunction,
  dropdownButton?: StyleFunction,
  displayedValue?: StyleFunction,
  dropdownWrapper?: StyleFunction,
  groupContainer?: StyleFunction,
  groupHeading?: StyleFunction,
  optionContainer?: StyleFunction,
  optionItem?: StyleFunction,
};

export interface DropdownState {
  open: boolean,
  searchTerm: string,
  searchTimer: NodeJS.Timer | null,
  focusedIndex: number,
  internalSelectedOption: string,
};

export interface DropdownProps {
  ariaDescribedBy: string,
  ariaLabel: string,
  ariaLabelledBy: string,
  arrowRenderer: (open: boolean) => React.ReactNode,
  buttonClassName: string,
  centerText: boolean,
  contentClassName: string,
  disabled: boolean,
  height: number,
  hideArrow: boolean,
  id: string,
  maxContentHeight: number,
  openUp: boolean,
  options: DropdownOption[],
  optionClassName: string,
  optionRenderer: (selectedOption: string, optionsArray: DropdownOption[], onOptionClicked: (e:  MouseEvent<HTMLButtonElement> | KeyboardEvent<HTMLButtonElement>) => void, elementsRef: any[], getStyle: (key: string, extraState: {}) => string) => React.ReactNode,
  pageKeyTraverseSize: number,
  placeholder: string,
  searchable: boolean,
  selectedOption: string,
  selectedOptionClassName: string,
  selectedValueClassName: string,
  setSelected: (option: string) => void,
  style: DropdownStyle,
  width: number,
}

class Dropdown extends Component<DropdownProps, DropdownState> {
  public static defaultProps = {
    ariaDescribedBy: null,
    ariaLabel: null,
    ariaLabelledBy: null,
    arrowRenderer: undefined,
    buttonClassName: undefined,
    centerText: false,
    contentClassName: null,
    disabled: false,
    height: null,
    hideArrow: false,
    id: null,
    maxContentHeight: null,
    openUp: false,
    optionClassName: null,
    optionRenderer: undefined,
    options: [],
    pageKeyTraverseSize: 10,
    placeholder: 'Select ...',
    searchable: true,
    selectedOption: null,
    selectedOptionClassName: null,
    selectedValueClassName: null,
    style: {},
    width: null,
  };

  public state: DropdownState;
  private elements: HTMLButtonElement[] = [];
  private container = createRef<HTMLDivElement>();
  private button = createRef<HTMLButtonElement>();

  constructor(props: DropdownProps) {
    super(props);

    let defaultOption = props.options[0] || { value: ''};
    if (defaultOption && (defaultOption as OptionGroup).groupOptions) {
      defaultOption = (defaultOption as OptionGroup).groupOptions[0];
    }

    this.state = {
      focusedIndex: -1,
      internalSelectedOption: (defaultOption as Option).value,
      open: false,
      searchTerm: '',
      searchTimer: null,
    };
  }

  public componentDidMount() {
    document.addEventListener('mouseup', this.onClick, false); // eslint-disable-line no-undef
    document.addEventListener('touchend', this.onClick, false); // eslint-disable-line no-undef
  }

  public componentWillUnmount() {
    document.removeEventListener('mouseup', this.onClick); // eslint-disable-line no-undef
    document.removeEventListener('touchend', this.onClick); // eslint-disable-line no-undef
  }

  public render() {
    // Please Keep Alphabetical
    const {
      ariaDescribedBy,
      ariaLabel,
      ariaLabelledBy,
      arrowRenderer,
      contentClassName,
      buttonClassName,
      disabled,
      hideArrow,
      id,
      placeholder,
      selectedOption,
      selectedValueClassName,
    } = this.props;

    const {
      internalSelectedOption,
      open,
    } = this.state;

    const displayedValue = selectedOption || internalSelectedOption || placeholder || '';
    const wrapperClass = this.getStyle('dropdownWrapper');
    const dropdownButtonClass = cx(buttonClassName, this.getStyle('dropdownButton'));
    const displayedValueClass = cx(selectedValueClassName, this.getStyle('displayedValue'));
    const contentClass = cx(contentClassName, this.getStyle('optionContainer'));
    const arrowClass = this.getStyle('arrow');

    return (
      <div
        className={wrapperClass}
        onKeyDown={this.onKeyDown}
        ref={this.container}
      >
        <button
          aria-label={ariaLabel}
          ari-describedby={ariaDescribedBy}
          aria-labelledby={ariaLabelledBy}
          className={dropdownButtonClass}
          disabled={disabled}
          id={id}
          onClick={this.onDropdownClick}
          onKeyDown={this.onDropdownClick}
          ref={this.button}
          type="button"
        >
          <div className={displayedValueClass}>{ displayedValue }</div>
          { !hideArrow && !arrowRenderer && <div className={arrowClass} /> }
          { !hideArrow && arrowRenderer && arrowRenderer(open) }
        </button>
        <ul className={contentClass}>{ this.renderOptions() }</ul>
      </div>
    );
  }

  private renderOptions = () => {
    const { optionRenderer, selectedOption, selectedOptionClassName, optionClassName, options } = this.props;
    const { internalSelectedOption } = this.state;
    this.elements = []; // Reset ref array

    if (optionRenderer) {
      return optionRenderer(selectedOption || internalSelectedOption, options, this.onOptionClicked, this.elements, this.getStyle);
    }

    return defaultOptionRenderer(selectedOption, options, selectedOptionClassName, optionClassName, this.onOptionClicked, this.elements, this.getStyle);
  }

  private onClick = (e: Event) => {
    if (this.container.current && !this.container.current.contains(e.target as Node)) {
      this.closeDropdown();
    }
  }

  private onDropdownClick = ({ nativeEvent }: MouseEvent<HTMLButtonElement> | KeyboardEvent<HTMLButtonElement>) => {
    const { disabled } = this.props;

    if (nativeEvent instanceof KeyboardEvent) { // eslint-disable-line no-undef
      if (nativeEvent.keyCode !== KEY_CODES.ENTER) return;
      nativeEvent.preventDefault();
    }

    if (!disabled) {
      this.setState((p: DropdownState) => ({ open: !p.open, focusedIndex: p.open ? -1 : p.focusedIndex }));
    }
  }

  private onOptionClicked = ({ nativeEvent }: MouseEvent<HTMLButtonElement> | KeyboardEvent<HTMLButtonElement>) => {
    const { setSelected } = this.props;

    if (nativeEvent instanceof KeyboardEvent) {
      if (nativeEvent.keyCode !== KEY_CODES.ENTER) return;
      nativeEvent.preventDefault();
    }

    if (nativeEvent.target) {
      const selectedOption = (nativeEvent.target as HTMLButtonElement).innerText;
      setSelected(selectedOption);
      this.setState({ open: false, internalSelectedOption: selectedOption });
      if (nativeEvent instanceof KeyboardEvent && nativeEvent.keyCode && nativeEvent.keyCode === KEY_CODES.ENTER && this.button.current) {
        this.button.current.focus();
      }
    }
  }

  private onKeyDown = ({ nativeEvent }: KeyboardEvent) => {
    const key = nativeEvent.key && nativeEvent.key.toLowerCase();
    const { keyCode } = nativeEvent;
    const { searchable } = this.props;

    if (NAVIGATION_KEYS.indexOf(keyCode) !== -1) {
      nativeEvent.preventDefault();
      this.onNavigation(keyCode);
    } else if (keyCode === KEY_CODES.TAB) {
      this.closeDropdown();
    } else if (key.length === 1 && searchable) {
      this.searchDropdown(key);
    }
  }

  private onNavigation = (keyCode: number) => {
    const { focusedIndex } = this.state;
    const { pageKeyTraverseSize } = this.props;

    switch (keyCode) {
      case KEY_CODES.UP_ARROW:
        if (focusedIndex === -1) {
          this.setState({ focusedIndex: 0 }, this.setFocus);
        } else if (focusedIndex === 0) {
          this.setState({ focusedIndex: this.elements.length - 1 }, this.setFocus);
        } else {
          this.setState(p => ({ focusedIndex: p.focusedIndex - 1 }), this.setFocus);
        }
        break;
      case KEY_CODES.DOWN_ARROW:
        this.setState(p => ({ focusedIndex: (p.focusedIndex + 1) % this.elements.length }), this.setFocus);
        break;
      case KEY_CODES.PAGE_UP:
        if (focusedIndex === -1) {
          this.setState({ focusedIndex: 0 }, this.setFocus);
        } else if (focusedIndex - pageKeyTraverseSize < 0) {
          this.setState({ focusedIndex: this.elements.length - 1 }, this.setFocus);
        } else {
          this.setState(p => ({ focusedIndex: p.focusedIndex - pageKeyTraverseSize - 1 }), this.setFocus);
        }
        break;
      case KEY_CODES.PAGE_DOWN:
        this.setState(p => ({ focusedIndex: (p.focusedIndex + pageKeyTraverseSize - 1) % this.elements.length }), this.setFocus);
        break;
      case KEY_CODES.ESCAPE:
        this.closeDropdown(true);
        break;
      default:
        break;
    }
  }

  private getStyle = (key: string, extraState?: {}) => {
    const { style } = this.props;
    const baseStyle = defaultStyles[key](this.props, this.state, extraState);
    const customStyle = style[key];
    return customStyle ? css(customStyle(baseStyle, this.state, extraState)) : css(baseStyle);
  }

  private setFocus = () => {
    const { focusedIndex } = this.state;
    const element = this.elements[focusedIndex];
    if (element) {
      element.focus();
    }
  }

  private closeDropdown = (focus = false) => {
    this.setState(p => ({ open: false, focusedIndex: p.internalSelectedOption ? p.focusedIndex : -1 }));
    if (focus && this.button.current) {
      this.button.current.focus();
    }
  }

  private searchDropdown = (key: string) => {
    const { searchTerm } = this.state;
    const oldTerm = searchTerm;
    this.setState(p => ({ searchTerm: p.searchTerm + key }));
    this.searchList(oldTerm + key);

    this.clearTimer();
    const timer = setTimeout(this.clearSearch, 1500);
    this.setState({ searchTimer: timer });
  }

  private clearTimer = () => {
    const { searchTimer } = this.state;
    if (searchTimer) {
      clearTimeout(searchTimer);
      this.setState({ searchTimer: null });
    }
  }

  private clearSearch = () => {
    this.setState({ searchTerm: '' });
  }

  private searchList = (value: string) => {
    const element: React.RefObject<HTMLButtonElement> = (this.elements as any).find(
      (el: React.RefObject<HTMLButtonElement>) => el.current && el.current.innerText.toLowerCase().indexOf(value) === 0,
    );
    if (element.current) element.current.focus();
  }
}

export default Dropdown;
