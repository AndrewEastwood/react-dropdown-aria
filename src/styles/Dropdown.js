import colours from './Colours';
import { optionStyle } from '../components/OptionItem';

const dropdownWrapperStyle = ({ width, height }) => ({
  width,
  height,
  position: 'relative',
  display: 'flex',
  flexDirection: 'column',
});

const dropdownButtonStyle = () => ({
  fontSize: '1em',
  display: 'flex',
  flexDirection: 'row',
  alignItems: 'center',
  width: '100%',
  height: '100%',
  backgroundColor: 'white',
  padding: '9px 5px',
  margin: '0',
  borderRadius: '0',
  borderBottom: `2px solid ${colours.greys.light}`,
  borderTop: 'none',
  borderRight: 'none',
  borderLeft: 'none',
  textAlign: 'left',
  cursor: 'pointer',
  outline: 'none',

  '&:hover': {
    boxShadow: `0px 1px 3px 2px ${colours.greys.lighter}`,
  },

  '&:focus': {
    boxShadow: `0px 1px 3px 2px ${colours.greys.lighter}`,
  },

  '&:disabled': {
    cursor: 'not-allowed',
  },
});

const displayedValueStyle = ({ hideArrow, selectedOption, centerText }, { internalSelectedOption }) => ({
  flex: '1',
  borderRight: hideArrow ? 'none' : `1px solid ${colours.greys.light}`,
  overflow: 'hidden',
  textOverflow: 'ellipsis',
  whiteSpace: 'nowrap',
  color: (selectedOption || internalSelectedOption) ? 'black' : colours.greys.base,
  textAlign: centerText ? 'center' : 'left',
});

const arrowStyle = (props, { open }) => ({
  content: '""',
  width: '0',
  height: '0',
  marginLeft: '8px',
  marginRight: '5px',
  borderRight: '5px solid transparent',
  borderLeft: '5px solid transparent',
  borderTop: open ? '0' : `5px solid ${colours.greys.base}`,
  borderBottom: open ? `5px solid ${colours.greys.base}` : '0',
});

const optionContainerStyle = ({ openUp, maxContentHeight }, { open }) => ({
  width: '100%',
  maxHeight: maxContentHeight || '175px',
  overflowY: maxContentHeight ? 'scroll' : null,
  zIndex: '9999',
  overflowX: 'hidden',
  position: 'absolute',
  left: '0',
  listStyleType: 'none',
  margin: '0',
  padding: '2px 0',
  backgroundColor: 'white',
  color: 'black',
  borderRadius: '2px',
  display: open ? 'block' : 'none',
  boxSizing: 'border-box',
  top: openUp ? null : '100%',
  bottom: openUp ? '105%' : null,
  boxShadow: openUp ? `0px -3px 3px 2px ${colours.greys.lighter}` : `0px 3px 3px 2px ${colours.greys.lighter}`,
});

const groupContainerStyle = () => ({
  padding: '1em 0 0 0',
});

const groupHeadingStyle = () => ({
  color: 'grey',
  fontSize: '0.9em',
  padding: '0 10px 3px 5px',
  display: 'flex',
  flexDirection: 'row',
  justifyContent: 'space-between',
});

const defaultStyles = {
  arrowStyle,
  dropdownButtonStyle,
  displayedValueStyle,
  dropdownWrapperStyle,
  groupContainerStyle,
  groupHeadingStyle,
  optionContainerStyle,
  optionStyle,
};

export default defaultStyles;
