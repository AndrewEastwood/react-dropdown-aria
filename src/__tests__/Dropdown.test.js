import React from 'react';
import { shallow, mount } from 'enzyme';
import sinon from 'sinon';
import toJson from 'enzyme-to-json';
import { createSerializer } from 'jest-emotion';
import * as emotion from 'emotion';
import Dropdown from '../Dropdown';
import { CUSTOM_OPTIONS, OPTIONS, GROUPED_OPTIONS } from './constants';
import { KEY_CODES } from '../utils/helper';

expect.addSnapshotSerializer(createSerializer(emotion));

const foo = () => {};

describe('Check Props', () => {
  it('Matches snapshot with default props', () => {
    const dropdown = shallow(<Dropdown setSelected={foo} />);
    expect(toJson(dropdown)).toMatchSnapshot();
  });

  it('Matches snapshot with custom props', () => {
    const dropdown = shallow(
      <Dropdown
        placeholder="Custom Placeholder..."
        buttonClassName="custom-class"
        id="dropdown"
        ariaLabel="React Simple Dropdown"
        options={CUSTOM_OPTIONS}
        setSelected={foo}
        disabled
        width={400}
        maxContentHeight={150}
      />,
    );
    expect(toJson(dropdown)).toMatchSnapshot();
  });

  it('Matches snapshot with grouped options', () => {
    const dropdown = shallow(<Dropdown setSelected={foo} options={GROUPED_OPTIONS} />);
    expect(toJson(dropdown)).toMatchSnapshot();
  });
});

describe('Navigation', () => {
  let wrapper;
  let button;

  beforeEach(() => {
    wrapper = mount(<Dropdown options={OPTIONS} setSelected={foo} buttonClassName="test" />);
    button = wrapper.find('.test'); // Actual dropdown button element
  });

  it('Opens dropdown when clicked', () => {
    button.simulate('click');

    expect(wrapper.state('open')).toBeTruthy();
  });

  it('Opens dropdown with enter key', () => {
    button.simulate('keyDown', { nativeEvent: { key: 'enter', keyCode: KEY_CODES.ENTER, preventDefault: foo } });

    expect(wrapper.state('open')).toBeTruthy();
  });

  it('Closes when clicked again', () => {
    button.simulate('click');
    expect(wrapper.state('open')).toBeTruthy();

    button.simulate('click');
    expect(wrapper.state('open')).toBeFalsy();
  });

  it('Closes when enter pressed again', () => {
    button.simulate('keyDown', { nativeEvent: { key: 'enter', keyCode: KEY_CODES.ENTER, preventDefault: foo } });
    expect(wrapper.state('open')).toBeTruthy();

    button.simulate('keyDown', { nativeEvent: { key: 'enter', keyCode: KEY_CODES.ENTER, preventDefault: foo } });
    expect(wrapper.state('open')).toBeFalsy();
  });

  it('Closes when tab key pressed', () => {
    button.simulate('click');
    expect(wrapper.state('open')).toBeTruthy();

    button.simulate('keyDown', { nativeEvent: { key: 'tab', keyCode: KEY_CODES.TAB } });
    expect(wrapper.state('open')).toBeFalsy();
  });

  it('Closes when esc pressed', () => {
    button.simulate('click');
    expect(wrapper.state('open')).toBeTruthy();

    button.simulate('keyDown', { nativeEvent: { key: 'escape', keyCode: KEY_CODES.ESCAPE, preventDefault: foo } });
    expect(wrapper.state('open')).toBeFalsy();
  });

  it('Arrow key selects first element in list', () => {
    button.simulate('click');
    wrapper.simulate('keyDown', { nativeEvent: { key: 'downArrow', keyCode: KEY_CODES.DOWN_ARROW, preventDefault: foo } });

    expect(document.activeElement.innerHTML).toBe('1'); // eslint-disable-line no-undef
  });

  it('Arrow key nav loops arround', () => {
    button.simulate('click');
    for (let i = 0; i < 10; i += 1) {
      wrapper.simulate('keyDown', { nativeEvent: { key: 'downArrow', keyCode: KEY_CODES.DOWN_ARROW, preventDefault: foo } });
    }
    expect(document.activeElement.innerHTML).toBe('10'); // eslint-disable-line no-undef

    wrapper.simulate('keyDown', { nativeEvent: { key: 'downArrow', keyCode: KEY_CODES.DOWN_ARROW, preventDefault: foo } });
    expect(document.activeElement.innerHTML).toBe('1'); // eslint-disable-line no-undef
  });
});

describe('Selecting Options', () => {
  let wrapper;
  let button;
  let spy;

  beforeEach(() => {
    spy = sinon.spy();
    wrapper = mount(<Dropdown options={OPTIONS} setSelected={spy} buttonClassName="test" />);
    button = wrapper.find('.test'); // Actual dropdown button element
  });

  it('calls setSelected with value when option selected', () => {
    button.simulate('click');
    wrapper.find('ul').childAt(0).simulate('click', { nativeEvent: { target: { innerText: '1' } } });

    expect(spy.calledWith('1')).toBe(true);
  });

  it('sets dropdown text to selected item', () => {
    button.simulate('click');
    wrapper.find('ul').childAt(0).simulate('click', { nativeEvent: { target: { innerText: '1' } } });

    expect(button.find('div').first().text()).toBe('1');
  });
});

describe('Special props', () => {
  it('Does not open when disabled', () => {
    const wrapper = mount(<Dropdown options={OPTIONS} setSelected={foo} buttonClassName="test" disabled />);
    wrapper.find('.test').simulate('click');

    expect(wrapper.state('open')).toBeFalsy();
  });
});
