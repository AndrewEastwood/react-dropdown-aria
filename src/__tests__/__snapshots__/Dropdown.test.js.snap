// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check Props Matches snapshot with custom props 1`] = `
<div
  className="dropdown"
  onKeyDown={[Function]}
  style={
    Object {
      "height": null,
      "width": 400,
    }
  }
>
  <button
    ari-describedby={null}
    aria-label="React Simple Dropdown"
    aria-labelledby={null}
    className="dropdown-select custom-class"
    disabled={true}
    id="dropdown"
    onClick={[Function]}
    type="button"
  >
    <div
      className="displayed-value grey"
    >
      Custom Placeholder...
    </div>
    <div
      className="dropdown-arrow down"
    />
  </button>
  <ul
    className="dropdown-content dropdown-content-down"
    style={
      Object {
        "maxHeight": 150,
        "overflowY": "scroll",
      }
    }
  >
    <button
      aria-label="ariaLabel"
      className="className dropdown-option"
      key="1"
      onClick={[Function]}
      tabIndex="-1"
      title="title"
      type="button"
    >
      <i
        className="fab fa-js-square option-icon"
      />
      1
    </button>
    <button
      aria-label="ariaLabel"
      className="className dropdown-option"
      key="2"
      onClick={[Function]}
      tabIndex="-1"
      title="title"
      type="button"
    >
      <i
        className="fab fa-js-square option-icon"
      />
      2
    </button>
    <button
      aria-label="ariaLabel"
      className="className dropdown-option"
      key="3"
      onClick={[Function]}
      tabIndex="-1"
      title="title"
      type="button"
    >
      <i
        className="fab fa-js-square option-icon"
      />
      3
    </button>
  </ul>
</div>
`;

exports[`Check Props Matches snapshot with default props 1`] = `
<div
  className="dropdown"
  onKeyDown={[Function]}
  style={
    Object {
      "height": null,
      "width": null,
    }
  }
>
  <button
    ari-describedby={null}
    aria-label={null}
    aria-labelledby={null}
    className="dropdown-select"
    disabled={false}
    onClick={[Function]}
    type="button"
  >
    <div
      className="displayed-value grey"
    >
      Select ...
    </div>
    <div
      className="dropdown-arrow down"
    />
  </button>
  <ul
    className="dropdown-content dropdown-content-down"
    style={Object {}}
  />
</div>
`;
